### DISCUSSION TITLE
Integrating Socket.IO with the App Router

### DISCUSSION THREAD

**Discussion by lucaspiresfernandes** [vercel/next.js#50097] [App Router]
Hello! Has anyone been successful at integrating the Socket.IO server with the app router?

Using the pages folder, it is quite easy to bind the Websocket server to `res.socket.server.io`, but I don't see a way to do this now that Route Handlers don't provide the `res` object.

Can anyone help me to figure this out?

**Comment by laureanray** [comment]
I don't get why is this closed? Can't even find the related discussion/issue.

**Reply by lucaspiresfernandes** [reply]
This is a duplicate post, so I figured I should just close it. This one exposes the problem much better than me: https://github.com/vercel/next.js/discussions/47782

**Comment by masadamsahid** [comment]
Please check out the example provided by SocketIO here: https://socket.io/how-to/use-with-nextjs

It seems we have to create a custom server for our Next.js project that combining the Next.js app handler and other node.js server (like using something like `http` or `express` server). Read more about custom server: 
- [Official Next.js docs, using `http` module](https://nextjs.org/docs/pages/building-your-application/configuring/custom-server)
- [Example of combining Next.js handlers to Express](https://github.com/vercel/next.js/issues/12656#issuecomment-641025110)

**Comment by Kamleshpaul** [comment]
For anyone looking for a detailed guide on using WebSocket in Next.js with the app router and Socket.IO, I’ve written an article that might help: [How to Use WebSocket in Next.js App Router with Socket.IO](https://codingtricks.co/posts/how-to-use-websocket-in-nextjs-app-router-with-socketio).  

The article covers:  
- Setting up WebSocket with Next.js App Router.  
- Integrating Socket.IO for real-time communication.  
- Handling server-side and client-side connections effectively.  

I hope this can save you some time and effort. Feedback and suggestions are always welcome!

**Reply by FurkanCodes** [reply]
Very  good article. As far as I understand, even though we host WebSocket server somewhere else the moment we add WebSocket to our nextjs client-side app, deploying it on vercel is out the window right?

**Reply by Kamleshpaul** [reply]
yes websocket server can be anywhere
